pkg/controller/cert_control.go

```diff
@@ -0,0 +1,289 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"encoding/pem"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	certutil "github.com/pingcap/tidb-operator/pkg/util/crypto"
+	capi "k8s.io/api/certificates/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	types "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+)
+
+// CertControlInterface manages certificates used by TiDB clusters
+type CertControlInterface interface {
+	Create(tc *v1alpha1.TidbCluster, commonName string, hostList []string, IPList []string, suffix string) error
+	LoadFromSecret(ns string, secretName string) ([]byte, []byte, error)
+	SaveToSecret(ns string, instance string, component string, cert []byte, key []byte) error
+	CheckSecret(ns string, secretName string) bool
+	//RevokeCert() error
+	//RenewCert() error
+}
+
+type realCertControl struct {
+	kubeCli kubernetes.Interface
+}
+
+// NewRealCertControl creates a new CertControlInterface
+func NewRealCertControl(
+	kubeCli kubernetes.Interface,
+) CertControlInterface {
+	return &realCertControl{
+		kubeCli: kubeCli,
+	}
+}
+
+func (rcc *realCertControl) Create(tc *v1alpha1.TidbCluster, commonName string,
+	hostList []string, IPList []string, suffix string) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+	csrName := fmt.Sprintf("%s-%s", tcName, suffix)
+
+	// generate certificate if not exist
+	_, _, err := rcc.LoadFromSecret(ns, csrName)
+	if !apierrors.IsNotFound(err) {
```

Comment:
The above comment seems wrong now. I would annotate this error with "secret csrName not found". `err != nil` is not handled now